rustls/src/crypto/mod.rs

```diff
@@ -484,6 +484,87 @@ pub trait ActiveKeyExchange: Send + Sync {
+    /// and the other is "classical".  See
+    /// <https://datatracker.ietf.org/doc/draft-ietf-tls-hybrid-design/11/>
+    ///
+    /// # Background
+    /// Rustls always sends a presumptive key share in its `ClientHello`, using
+    /// (absent any other information) the first item in [`CryptoProvider::kx_groups`].
+    /// If the server accepts the client's selection, it can complete the handshake
+    /// using that key share.  If not, the server sends a `HelloRetryRequest` instructing
+    /// the client to send a different key share instead.
+    ///
+    /// This request costs an extra round trip, and wastes the key exchange computation
+    /// (in [`SupportedKxGroup::start()`]) the client already did.  We would
+    /// like to avoid those wastes if possible.
+    ///
+    /// It is early days for post-quantum-secure hybrid key exchange deployment.
+    /// This means (commonly) continuing to offer both the hybrid and classical
+    /// key exchanges, so the handshake can be completed without a `HelloRetryRequest`
+    /// for servers that support the offered hybrid or classical schemes.
+    ///
+    /// Implementing `hybrid_component()` enables two optimizations:
+    ///
+    /// 1. Sending both the hybrid and classical key shares in the `ClientHello`.
+    ///
+    /// 2. Performing the classical key exchange setup only once.  This is important
+    ///    because the classical key exchange setup is relatively expensive.
+    ///    This optimization is permitted and described in
+    ///    <https://www.ietf.org/archive/id/draft-ietf-tls-hybrid-design-11.html#section-3.2>
+    ///
+    /// Both of these only happen if the classical algorithm appears separately in
+    /// the client's [`CryptoProvider::kx_groups`], and if the hybrid algorithm appears
+    /// first in that list.
+    ///
+    /// # How it works
+    /// This function is only called by rustls for clients.  It is called when
+    /// constructing the initial `ClientHello`.  rustls follows these steps:
+    ///
+    /// 1. If the return value is `None`, nothing further happens.
+    /// 2. If the given [`NamedGroup`] does not appear in
+    ///    [`CryptoProvider::kx_groups`], nothing further happens.
+    /// 3. The given key share is added to the `ClientHello`, after the hybrid entry.
+    ///
+    /// Then, one of three things may happen when the server replies to the `ClientHello`:
+    ///
+    /// 1. The server sends a `HelloRetryRequest`.  Everything is thrown away and
+    ///    we start again.
+    /// 2. The server agrees to our hybrid key exchange: rustls calls
+    ///    [`ActiveKeyExchange::complete()`] consuming `self`.
+    /// 3. The server agrees to our classical key exchange: rustls calls
+    ///    [`ActiveKeyExchange::complete_hybrid_component()`] which
+    ///    discards the hybrid key data, and completes just the classical key exchange.
+    fn hybrid_component(&self) -> Option<(NamedGroup, &[u8])> {
+        None
+    }
```

Comment:
Still noodling on this, and the `unreachable!()` in `complete_hybrid_component()`... Why couldn't we yield an `enum` here that contains either `Self` or some new type (somewhat like `KeyExchangeChoice`) that represents the ability to do a hybrid key exchange?

Reply:
I think the receiver would need to be `self: Box<Self>` or else the impls of it need to do run-time typing to get their state back, but that does mean you can't go back from that to `Box<dyn ActiveKeyExchange>` in the default impl.

Obviously passing the original `Box<dyn ActiveKeyExchange>` alongside a normal receiver (either the receiver borrows the object before it is moved into the argument, or it is moved too and the parameter can't be moved a second time.)

Reply:
@djc @cpu any further thoughts on this?

Reply:
I think I'm comfortable with the approach in-branch and haven't come up with any workable ideas that move the needle.

Reply:
Going to give it one more think tomorrow morning when I'm fresh.

Reply:
So started noodling around with this and still wondering whether it's possible to figure this out:

```rust
    fn hybrid_component(self: Box<Self>) -> MaybeHybrid {
        MaybeHybrid::Single(Box::new(self))
    }

enum MaybeHybrid {
    Single(Box<dyn ActiveKeyExchange>),
}
```

This doesn't work directly:

```rust
error[E0277]: the size for values of type `Self` cannot be known at compilation time
   --> rustls/src/crypto/mod.rs:546:29
    |
546 |         MaybeHybrid::Single(Box::new(self))
    |                             ^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
note: required for `alloc::boxed::Box<Self>` to implement `crypto::ActiveKeyExchange`
   --> rustls/src/crypto/mod.rs:596:28
    |
596 | impl<T: ActiveKeyExchange> ActiveKeyExchange for Box<T> {
    |      -                     ^^^^^^^^^^^^^^^^^     ^^^^^^
    |      |
    |      unsatisfied trait bound introduced here
    = note: required for the cast from `alloc::boxed::Box<alloc::boxed::Box<Self>>` to `alloc::boxed::Box<(dyn crypto::ActiveKeyExchange + 'static)>`
help: consider further restricting `Self`
    |
545 |     fn hybrid_component(self: Box<Self>) -> MaybeHybrid where Self: Sized {
    |                                                         +++++++++++++++++
```

But I'm wondering if there's a way out of this using either an extra trait or maybe an extra layer of boxing?

Reply:
Right, so what about this?

https://github.com/rustls/rustls/commit/71edf410ac5842971dbe5904c03f247ce83f17d1

Reply:
doesn't seem to compile? 

```
error: reached the recursion limit while instantiating `<alloc::boxed::Box<alloc::boxed::Box<alloc::boxed::Box<alloc::boxed::Box<alloc::boxed::Box<...>>>>> as crypto::ActiveKeyExchange>::maybe_hybrid`
   --> rustls/src/crypto/mod.rs:550:29
```

i wonder if we can put the `maybe_hybrid` fn in its own trait, and then have a blanket impl of that trait for any type that impls `ActiveKeyExchange`?

Reply:
> i wonder if we can put the `maybe_hybrid` fn in its own trait, and then have a blanket impl of that trait for any type that impls `ActiveKeyExchange`?

this breaks coherence